File: BowlerBuilder/KotlinUI/src/main/kotlin/com/neuronrobotics/bowlerbuilder/view/cad/cadengine/camera/VirtualCameraDevice.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.neuronrobotics.bowlerbuilder.view.cad.cadengine.camera;

import com.neuronrobotics.bowlerstudio.physics.TransformFactory;
import com.neuronrobotics.imageprovider.AbstractImageProvider;
import com.neuronrobotics.sdk.addons.kinematics.math.RotationNR;
import com.neuronrobotics.sdk.addons.kinematics.math.TransformNR;
import java.awt.image.BufferedImage;
import java.util.ArrayList;
import javafx.scene.Camera;
import javafx.scene.Group;
import javafx.scene.transform.Affine;
import javax.annotation.ParametersAreNonnullByDefault;

@ParametersAreNonnullByDefault
public class VirtualCameraDevice extends AbstractImageProvider {

  private static final int DEFAULT_ZOOM_DEPTH = -1500;
  private static final Affine OFFSET =
      TransformFactory.nrToAffine(new TransformNR(0, 0, 0, new RotationNR(180, 0, 0)));
  private final Group cameraFrame = new Group();
  private Camera camera;
  private double zoomDepth = getDefaultZoomDepth();
  private final Affine zoomAffine = new Affine();
  private final Group manipulationFrame;

  public VirtualCameraDevice(final Camera camera, final Group hand) {
    super();
    setCamera(camera);
    setScriptingName("virtualCameraDevice");

    manipulationFrame = new Group();
    camera.getTransforms().add(zoomAffine);

    cameraFrame.getTransforms().add(getOffset());
    manipulationFrame.getChildren().addAll(camera, hand);
    cameraFrame.getChildren().add(manipulationFrame);

    setZoomDepth(DEFAULT_ZOOM_DEPTH);
  }

  public static int getDefaultZoomDepth() {
    return DEFAULT_ZOOM_DEPTH;
  }

  public static Affine getOffset() {
    return OFFSET;
  }

  @Override
  public void setGlobalPositionListener(final Affine affine) {
    super.setGlobalPositionListener(affine);
    manipulationFrame.getTransforms().clear();
    manipulationFrame.getTransforms().add(affine);
  }

  @Override
  protected boolean captureNewImage(final BufferedImage imageData) {
    return false;
  }

  @Override
  public BufferedImage captureNewImage() {
    return null;
  }

  /** Nothing to disconnect. */
  @Override
  public void disconnectDeviceImp() {
    // Not used
  }

  /**
   * Nothing to connect.
   *
   * @return true
   */
  @Override
  public boolean connectDeviceImp() {
    return true;
  }

  @Override
  public ArrayList<String> getNamespacesImp() {
    return new ArrayList<>();
  }

  public Camera getCamera() {
    return camera;
  }

  private void setCamera(final Camera camera) {
    this.camera = camera;
  }

  public Group getCameraGroup() {
    return getCameraFrame();
  }

  public Group getCameraFrame() {
    return cameraFrame;
  }

  public double getZoomDepth() {
    return zoomDepth;
  }

  /**
   * Set the zoom depth to zoom the camera.
   *
   * @param zoomDepth new zoom depth
   */
  public void setZoomDepth(final double zoomDepth) {
    this.zoomDepth = zoomDepth;

    if (this.zoomDepth > -2) {
      this.zoomDepth = -2;
    }

    if (this.zoomDepth < -5000) {
      this.zoomDepth = -5000;
    }

    zoomAffine.setTz(this.zoomDepth);
  }
}

File: BowlerBuilder/KotlinUI/src/main/kotlin/com/neuronrobotics/bowlerbuilder/view/cad/cadengine/camera/VirtualCameraDevice.kt
/*
 * This file is part of BowlerBuilder.
 *
 * BowlerBuilder is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * BowlerBuilder is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with BowlerBuilder.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.neuronrobotics.bowlerbuilder.view.cad.cadengine.camera

import com.neuronrobotics.imageprovider.AbstractImageProvider
import com.neuronrobotics.sdk.addons.kinematics.TransformFactory
import com.neuronrobotics.sdk.addons.kinematics.math.RotationNR
import com.neuronrobotics.sdk.addons.kinematics.math.TransformNR
import javafx.scene.Camera
import javafx.scene.Group
import javafx.scene.transform.Affine
import java.awt.image.BufferedImage
import java.util.ArrayList

class VirtualCameraDevice(
    camera: Camera,
    hand: Group
) : AbstractImageProvider() {

    private val zoomAffine = Affine()
    private val manipulationFrame: Group

    val cameraFrame = Group()
    var zoomDepth = defaultZoomDepth.toDouble()
        set(newDepth) {
            field = newDepth

            if (this.zoomDepth > -2) {
                field = -2.0
            }

            if (this.zoomDepth < -5000) {
                field = -5000.0
            }

            zoomAffine.tz = this.zoomDepth
        }

    init {
        scriptingName = "virtualCameraDevice"

        manipulationFrame = Group()
        camera.transforms.add(zoomAffine)

        cameraFrame.transforms.add(offset)
        manipulationFrame.children.addAll(camera, hand)
        cameraFrame.children.add(manipulationFrame)
    }

    override fun setGlobalPositionListener(affine: Affine) {
        super.setGlobalPositionListener(affine)
        manipulationFrame.transforms.clear()
        manipulationFrame.transforms.add(affine)
    }

    override fun captureNewImage(imageData: BufferedImage): Boolean = false

    override fun captureNewImage(): BufferedImage? = null

    /** Nothing to disconnect.  */
    override fun disconnectDeviceImp() {
        // Not used
    }

    /**
     * Nothing to connect.
     *
     * @return true
     */
    override fun connectDeviceImp() = true

    override fun getNamespacesImp(): ArrayList<String> = ArrayList()

    companion object {
        const val defaultZoomDepth = -1500
        val offset: Affine = TransformFactory.nrToAffine(
            TransformNR(
                0.0,
                0.0,
                0.0,
                RotationNR(180.0, 0.0, 0.0)
            )
        )
    }
}
